### PR TITLE
Feature/linktag optional extra field

### LINKED ISSUE #5: TrustAtom has optional entry hash for extra attributes
# Feature

Example TA: `Ŧ→sushi[0x00].900000000[0x00]892412523[0x00]uhCEk…UFnFF`

TrustAtom create can take optional `extra` field which is stored as an Entry
containing a BTreeMap<String, String>

If that entry exists, the link tag  includes a string version (NOT raw bytes) of EntryHashB64 pointing to that entry.

# AC

- Tests: storing and retrieving key/value pairs

### PR DESCRIPTION
**Pull request by dauphin3**
fixes #5 and #18

### CODE REVIEW

File: zomes/hc_zome_trust_atom/src/lib.rs [hunk 1]
```diff
@@ -43,8 +43,9 @@ pub struct QueryMineInput {
 // ZOME API FUNCTIONS
 
 #[hdk_extern]
-pub fn create_trust_atom(input: TrustAtomInput) -> ExternResult<()> {
-  trust_atom::create(input.target, &input.content, &input.value, input.attributes)
+pub fn create_trust_atom(input: TrustAtomInput) -> ExternResult<TrustAtom> {
+  let trust_atom = ::create(input)?;
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/lib.rs:47]: let's go back to passing all the args individually to ::create(...)

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 2]
```diff
@@ -15,21 +16,20 @@ enum LinkDirection {
...
-  pub content: String,
-  pub value: String,
-  pub source_entry_hash: EntryHashB64,
-  pub target_entry_hash: EntryHashB64,
-  pub attributes: BTreeMap<String, String>,
+  pub source: EntryHash,
+  pub target: EntryHash, 
+  pub content: Option<String>,
+  pub value: Option<String>,
+  pub extra: Option<String>, // string version of EntryHash
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:23]:
```suggestion
  pub extra: Option<BTreeMap<String, String>>,
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 3]
```diff
@@ -15,21 +16,20 @@ enum LinkDirection {
...
 #[derive(Serialize, Deserialize, SerializedBytes, Debug, Clone, PartialEq)]
 pub struct TrustAtom {
-  pub source: String,
-  pub target: String,
-  pub content: String,
-  pub value: String,
-  pub source_entry_hash: EntryHashB64,
-  pub target_entry_hash: EntryHashB64,
-  pub attributes: BTreeMap<String, String>,
+  pub source: EntryHash,
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:19]:
```suggestion
  pub source: String,
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 4]
```diff
@@ -15,21 +16,20 @@ enum LinkDirection {
...
+  pub target: EntryHash,
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:20]:
```suggestion
  pub target: String, 
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 5]
```diff
@@ -112,14 +125,11 @@ fn trust_atom_link_tag(link_direction: &LinkDirection, mut chunks: Vec<&str>) ->
   let mut link_tag_bytes = vec![];
   link_tag_bytes.extend_from_slice(&LINK_TAG_HEADER);
   link_tag_bytes.extend_from_slice(&link_tag_arrow);
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:127]:
```suggestion
  link_tag_bytes.extend_from_slice(&link_tag_arrow);
  let content = chunks.remove(0);
  link_tag_bytes.extend_from_slice(content.as_bytes());  
```

File: zomes/hc_zome_trust_atom/src/lib.rs (hunk 1 above)
Comment (harlantwood) [reply zomes/hc_zome_trust_atom/src/lib.rs:47]: x

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 6]
```diff
@@ -15,95 +15,147 @@ enum LinkDirection {
...
+      _ => {}
+    };
+    if i == chunk_options.len() - 1 {
+      break; // do not add NUL character at the end
     }
-    Err(error) => Err(WasmError::Guest(format!(
-      "Value could not be processed: `{}`.  Error: `{}`",
-      value_str, error
-    ))),
+    chunks.push(UNICODE_NUL_STR.to_string());
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:152]:
```suggestion
```

File: zomes/hc_zome_trust_atom/src/lib.rs [hunk 7]
```diff
@@ -11,40 +11,53 @@
...
 use hdk::prelude::*;
+use std::collections::BTreeMap;
 
 // public for sweettest; TODO can we fix this:
 pub mod trust_atom;
 pub use crate::trust_atom::*;
-
 pub mod test_helpers;
-// pub use crate::test_helpers;
+pub use crate::test_helpers::*;
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/lib.rs:20]: Where do we use this?  Seems suspicious. 😎

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 8]
```diff
@@ -15,95 +15,143 @@ enum LinkDirection {
...
-  let forward_link_tag = trust_atom_link_tag(
-    &LinkDirection::Forward,
-    vec![content, &normalized_value, bucket.as_str()],
-  );
-  let reverse_link_tag = trust_atom_link_tag(
-    &LinkDirection::Reverse,
-    vec![content, &normalized_value, bucket.as_str()],
-  );
+  let extra_entry_hash = match extra.clone() {
+    Some(x) => Some(create_extra(x)?),
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:49]: Change x to something… extra?

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 9]
```diff
@@ -269,45 +314,28 @@ fn convert_link_to_trust_atom(
...
         source: link_base_b64.to_string(),
         target: link_target_b64.to_string(),
-        content,
-        value,
         source_entry_hash: link_base_b64,
         target_entry_hash: link_target_b64,
-        attributes: BTreeMap::new(), // TODO
+        content: Some(content),
+        value: Some(value),
+        extra: Some(BTreeMap::new()), // TODO
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:321]: We should populate this with extra from dht… And integration tests should check it. 

Maybe later we can make this optional.

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 10]
```diff
@@ -269,45 +314,28 @@ fn convert_link_to_trust_atom(
...
-        source_entry_hash: link_target_b64,
-        target_entry_hash: link_base.clone().into(),
-        attributes: BTreeMap::new(), // TODO
+        source: link_target_b64.to_string(), // flipped for Reverse direction
+        target: link_base_b64.to_string(),   // flipped for Reverse direction
+        source_entry_hash: link_target_b64,  // flipped for Reverse direction
+        target_entry_hash: link_base_b64,    // flipped for Reverse direction
+        content: Some(content),
+        value: Some(value),
+        extra: Some(BTreeMap::new()), // TODO
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:332]: Same as above

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 11]
```diff
@@ -151,15 +169,18 @@ pub async fn test_query_mine() {
...
       TrustAtomInput {
         target: target_entry_hash.clone(),
-        content: "sushi".into(),
-        value: "0.8".into(),
-        attributes: BTreeMap::new(),
+        content: Some("sushi".to_string()),
+        value: Some("0.8".to_string()),
+        extra: Some(BTreeMap::new()),
+        // extra: Some(BTreeMap::new([
+        //   ("creator_name".into(), "Bradley Fieldstone Jr.".into()),
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:182]: Either use commented out code or delete

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 12]
```diff
@@ -300,10 +322,10 @@ pub async fn test_query_mine_with_content_full() {
       "query_mine",
       QueryMineInput {
         target: None,
-        content_starts_with: None,
         content_full: Some("sushi".into()),
-        min_rating: None,
-        // min_rating: Some("0.0".into()),
+        content_starts_with: None,
+        value_starts_with: None,
+        // value_starts_with: Some("0.0".into()),
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:328]: Delete line

File: zomes/hc_zome_trust_atom/src/test_helpers.rs [hunk 13]
```diff
@@ -19,10 +33,50 @@ pub fn list_links(base: EntryHash, link_tag_text: Option<String>) -> ExternResul
...
+  let target_entry_hash = hash_entry(string_target)?;
+  Ok(target_entry_hash)
+}
 
 fn link_tag(tag: String) -> ExternResult<LinkTag> {
   let serialized_bytes: SerializedBytes = StringLinkTag(tag).try_into()?;
   Ok(LinkTag(serialized_bytes.bytes().clone()))
 }
+
+pub fn create_mock_extra_entry(input: BTreeMap<String, String>) -> ExternResult<EntryHash> {
```
[the remark this answers is not in the document]
Comment (dauphin3) [reply zomes/hc_zome_trust_atom/src/test_helpers.rs:50]: to create a "sweet" Extra struct entry with the example input extra data fields ie (k,v) "extra_stuff", "extra info here"

File: zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs [hunk 14]
```diff
@@ -353,3 +375,80 @@ pub async fn setup_conductors(n: usize) -> (SweetConductorBatch, Vec<AgentPubKey
...
+        "Say some extra stuff here".to_string(),
+      ),
+      (
+        "another_thing".to_string(),
+        "Put more information here".to_string(),
+      ),
+    ])),
+  };
+
+  let mock_trust_atom: TrustAtom = conductor
```
[the remark this answers is not in the document]
Comment (dauphin3) [reply zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:407]: that is a reasonable alternative I think, wanted to keep it modular, didn't feel as confident on the queries

[the remark this answers is not in the document]
Comment (dauphin3) [reply zomes/hc_zome_trust_atom/tests/trust_atom_tests.rs:407]: also this route could help set the stage for further test features (i think?)

File: zomes/hc_zome_trust_atom/src/lib.rs (hunk 7 above)
Comment (dauphin3) [reply zomes/hc_zome_trust_atom/src/lib.rs:20]: the test_helper functions at the bottom

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 15]
```diff
@@ -15,95 +15,147 @@ enum LinkDirection {
...
+    },
+    None => Ok(None),
+  }
+}
+
+fn create_link_tag(link_direction: &LinkDirection, chunk_options: Vec<Option<String>>) -> LinkTag {
+  let mut chunks: Vec<String> = vec![];
+
+  for i in 0..chunk_options.len() {
+    if let Some(_) = chunk_options[i].clone() {
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:146]:
```suggestion
    if let Some(chunk) = chunk_options[i].clone() {
```

File: zomes/hc_zome_trust_atom/src/trust_atom.rs [hunk 16]
```diff
@@ -15,95 +15,147 @@ enum LinkDirection {
...
+      chunks.push(chunk_options[i].clone().unwrap()); // unwrap should be fine since gaurded
```
Comment (harlantwood) [inline zomes/hc_zome_trust_atom/src/trust_atom.rs:147]:
```suggestion
      chunks.push(chunk); 
```

**Comment by harlantwood** [comment]
When you’re ready to merge, rebase to a single commit

File: zomes/hc_zome_trust_atom/src/trust_atom.rs (hunk 15 above)
Comment (dauphin3) [reply zomes/hc_zome_trust_atom/src/trust_atom.rs:146]: chunk would never get used in this block. no need to assign var

**Pull request merged**